odoo/orm/domains.py

```diff
@@ -1309,32 +1311,53 @@ def _optimize_type_datetime(condition, _):
     """Make sure we have a datetime type in the value"""
     if condition.operator.endswith('like') or "." in condition.field_expr:
         return condition
+    field_expr = condition.field_expr
     operator = condition.operator
     value, is_day = _value_to_datetime(condition.value)
-    if value is False and operator[0] in ('<', '>'):
-        # comparison to False results in an empty domain
-        return _FALSE_DOMAIN
-    if value == condition.value:
-        assert not is_day
-        return condition
 
-    # if we get a day we may need to add 1 depending on the operator
-    if is_day and operator == '>':
-        try:
-            value += timedelta(1)
-        except OverflowError:
-            # higher than max, not possible
+    # Handle inequalitiy
+    if operator[0] in ('<', '>'):
+        if value is False:
             return _FALSE_DOMAIN
-        operator = '>='
-    elif is_day and operator == '<=':
-        try:
-            value += timedelta(1)
-        except OverflowError:
-            # lower than max, just check if field is set
-            return DomainCondition(condition.field_expr, '!=', False)
-        operator = '<'
+        if not isinstance(value, datetime):
+            return condition
+        if value.microsecond:
+            assert not is_day, "days don't have microseconds"
+            value = value.replace(microsecond=0)
+        delta = timedelta(days=1) if is_day else timedelta(seconds=1)
+        if operator == '>':
+            try:
+                value += delta
+            except OverflowError:
+                # higher than max, not possible
+                return _FALSE_DOMAIN
+            operator = '>='
+        elif operator == '<=':
+            try:
+                value += delta
+            except OverflowError:
+                # lower than max, just check if field is set
+                return DomainCondition(field_expr, '!=', False)
+            operator = '<'
 
-    return DomainCondition(condition.field_expr, operator, value)
+    # Handle equality: compare to the whole second
+    if (
+        operator in ('in', 'not in')
+        and not is_day  # leave full days as-is for now
```

Comment:
Is this `and not is_day` necessary here?
I think after `_value_to_datetime`, each `v` in `value` is either `False` or `isinstance(v, datetime)` or `isinstance(v, SQL)`. So we can optimize `date` in `condition.value` which has been converted to `datetime` now, instead of leaving them to be optimized in the next around.